packages/appium/lib/appium.js

```diff
@@ -241,6 +251,261 @@ class AppiumDriver extends DriverCore {
+          );
+          // if we're meant to proxy to an upstream bidi socket, just do that
+          await sendToProxy(data.toString('utf8'));
+        } else {
+          const res = await this.onBidiMessage(data, bidiHandlerDriver);
+          await send(JSON.stringify(res));
+        }
+      });
+
+      // Next consider if the client closes the socket connection on us
+      ws.on('close', (code, reason) => {
+        // Not sure if we need to do anything here if the client closes the websocket connection.
+        // Probably if a session was started via the socket, and the socket closes, we should end the
+        // associated session to free up resources. But otherwise, for sockets attached to existing
+        // sessions, doing nothing is probably right.
+        this.log.debug(`Bidi socket connection closed (code ${code}, reason: '${reason}')`);
+
+        // If we're proxying, might as well close the upstream connection and clean it up
+        if (proxyClient) {
+          this.log.debug('Also closing bidi proxy socket connection');
+          proxyClient.close(code, reason);
+        }
+      });
+
+      // If the driver emits a bidi event that should maybe get sent to the client, check to make
+      // sure the client is subscribed and then pass it on
+      let eventListener = async ({context, method, params}) => {
+        // if the driver didn't specify a context, use the empty context
+        if (!context) {
+          context = '';
+        }
+        if (!method || !params) {
+          throw new Error(
+            `Driver emitted a bidi event that was malformed. Require method and params keys ` +
+              `(with optional context). But instead received: ${JSON.stringify({
+                context,
+                method,
+                params,
+              })}`,
+          );
+        }
+        if (ws.readyState !== 1) {
+          // if the websocket is not still 'open', then we can ignore sending these events
+          if (ws.readyState > 1) {
+            // if the websocket is closed or closing, we can remove this listener as well to avoid
+            // leaks
+            bidiHandlerDriver.eventEmitter.removeListener(BIDI_EVENT_NAME, eventListener);
+          }
+          return;
+        }
+
+        if (
+          bidiHandlerDriver.bidiEventSubs[method] &&
+          bidiHandlerDriver.bidiEventSubs[method].includes(context)
+        ) {
+          this.log.info(
+            `<-- BIDI EVENT ${method} (context: '${context}', params: ${JSON.stringify(params)})`,
+          );
+          // now we can send the event onto the socket
+          const ev = {type: 'event', context, method, params};
+          // TODO add conditional logic to send only if user is subscribed
+          await send(JSON.stringify(ev));
+        }
+      };
+      bidiHandlerDriver.eventEmitter.on(BIDI_EVENT_NAME, eventListener);
+    } catch (err) {
+      this.log.error(err);
+      try {
+        ws.close();
```

Comment:
can it ever throw an error?

Reply:
I can't tell from a brief look at the code. Do you think we should avoid the try/catch for now and then add it if errors ever occur? Seemed like a good idea to wrap third party library methods that we're not sure are safe. https://github.com/websockets/ws/blob/master/lib/websocket.js#L283